ambersim/trajopt/shooting.py

```diff
@@ -0,0 +1,169 @@
+        """
+        raise NotImplementedError
+
+
+# predictive sampling API
+
+
+@struct.dataclass
+class VanillaPredictiveSamplerParams(ShootingParams):
+    """Parameters for generic predictive sampling methods."""
+
+    key: jax.Array  # random key for sampling
+
+
+@struct.dataclass
+class VanillaPredictiveSampler(ShootingAlgorithm):
+    """A vanilla predictive sampler object.
+
+    The following choices are made:
+    (1) the model parameters are fixed, and are therefore a field of this dataclass;
+    (2) the control sequence is parameterized as a ZOH sequence instead of a spline;
+    (3) the control inputs are sampled from a normal distribution with a uniformly chosen noise scale over all params.
+    (4) the cost function is quadratic in the states and controls.
+    """
+
+    model: mjx.Model
+    cost_function: CostFunction
+    nsamples: int = struct.field(pytree_node=False)
+    stdev: float = struct.field(pytree_node=False)  # noise scale, parameters theta_new ~ N(theta, (stdev ** 2) * I)
+
+    def optimize(self, params: VanillaPredictiveSamplerParams) -> Tuple[jax.Array, jax.Array, jax.Array]:
+        """Optimizes a trajectory using a vanilla predictive sampler.
+
+        Args:
+            params: The parameters of the trajectory optimizer.
+
+        Returns:
+            qs (shape=(N + 1, nq)): The optimized trajectory.
+            vs (shape=(N + 1, nv)): The optimized generalized velocities.
+            us (shape=(N, nu)): The optimized controls.
+        """
+        # unpack the params
+        m = self.model
+        nsamples = self.nsamples
+        stdev = self.stdev
+
+        q0 = params.q0
+        v0 = params.v0
+        us_guess = params.us_guess
+        N = params.N
+        key = params.key
+
+        # sample over the control inputs
+        _us_samples = us_guess + jax.random.normal(key, shape=(nsamples, N, m.nu)) * stdev
+
+        # clamping the samples to their control limits
+        # TODO(ahl): write a create classmethod that allows the user to set default_limits optionally with some semi-
+        # reasonable default value
+        # TODO(ahl): check whether joints with no limits have reasonable defaults for m.actuator_ctrlrange
+        limits = m.actuator_ctrlrange
+        clip_fn = partial(jnp.clip, a_min=limits[:, 0], a_max=limits[:, 1])  # clipping function with limits already set
+        us_samples = vmap(vmap(clip_fn))(_us_samples)  # apply limits only to the last dim, need a nested vmap
+        # limited = m.actuator_ctrllimited[:, None]  # (nu, 1) whether each actuator has limited control authority
+        # default_limits = jnp.array([[-1000.0, 1000.0]] * m.nu)  # (nu, 2) default limits for each actuator
+        # limits = jnp.where(limited, m.actuator_ctrlrange, default_limits)  # (nu, 2)
+
+        # predict many samples, evaluate them, and return the best trajectory tuple
+        qs_samples, vs_samples = vmap(shoot, in_axes=(None, None, None, 0))(m, q0, v0, us_samples)
```

Comment:
Just from casually commenting stuff out, i've 100% confirmed that it's this `vmap` over `shoot` call which takes up nearly 100% of the runtime (not the stuff before or below it).